### PR TITLE
Season ticket

### PR DESCRIPTION
**Pull request by WBGroupR**

### CODE REVIEW

**Comment by WBGroupR** [comment]
Only Chris seems to be available for this review

**Comment by WBGroupR** [comment]
Now able to add Duane as a reviewer too.

File: src/bcccp/tickets/season/SeasonTicket.java [hunk 1]
```diff
@@ -18,61 +19,98 @@ public SeasonTicket (String ticketId,
...
+		boolean foundUsageRecord = false;
+
+		while (usageRec.hasNext()) {
+
+		    if (usageRec.next().getSeasonTicketId().equals(getId())) {
+
+                foundUsageRecord = true;
+            }
+
+            else foundUsageRecord = false;
```
Comment (duane-mcmahon) [inline src/bcccp/tickets/season/SeasonTicket.java:65]: 3. Type. Performance, General Programming:  break out of a loop if work is done.
Defect as per my own 'suboptimal' code in adhocticketdao. it's not something you necessarily remember to insert the break after 'hit' in a while loop when writing up the 1st draft. i would rework your method to:

@Override
  	public boolean inUse() {
	
 		boolean foundUsageRecord = false;
 
 		Iterator<IUsageRecord> usageRec = usages.iterator();
 
 		while (usageRec.hasNext()) {
 
 		    if (usageRec.next().getSeasonTicketId().equals(getId())) {
 
                 foundUsageRecord = true;

                 break;
           }

File: src/bcccp/tickets/season/SeasonTicket.java [hunk 2]
```diff
@@ -18,61 +19,98 @@ public SeasonTicket (String ticketId,
...
+
+        while (usageRecs.hasNext()) {
+
+            if (usageRecs.next().getSeasonTicketId().equals(getId())) {
+
+                currentUsage = usageRecs.next();
+            }
+        }
+
+        return currentUsage;
```
Comment (duane-mcmahon) [inline src/bcccp/tickets/season/SeasonTicket.java:90]: 4. Type. Performance, General Programming: break out of a loop if work is done.

File: src/bcccp/tickets/season/SeasonTicketDAO.java [hunk 3]
```diff
@@ -3,64 +3,108 @@
...
+
+		Iterator<ISeasonTicket> sTicketRecs = seasonTickets.iterator();
+
+		while (sTicketRecs.hasNext()) {
+
+			if (sTicketRecs.next().getId().equals(ticket.getId())) {
+
+				sTicketRecs.remove();
+			}
+		}
```
Comment (duane-mcmahon) [inline src/bcccp/tickets/season/SeasonTicketDAO.java:49]: 6. Type. Performance, General Programming: break out of a loop if work is done.

File: src/bcccp/tickets/season/SeasonTicketDAO.java [hunk 4]
```diff
@@ -3,64 +3,108 @@
...
+			if (sTicketRecs.next().getId().equals(ticketId)) {
+
+				sTicket = sTicketRecs.next();
+			}
+			else {
+				sTicket = null;
+			}
+		}
+		return sTicket;
+	}
```
Comment (duane-mcmahon) [inline src/bcccp/tickets/season/SeasonTicketDAO.java:79]: 7. Type. Performance, General Programming: break out of a loop if work is done.

8. Type. Exceptions: Don't return Null.
Question: is non-existence a perfectly valid return value for this function?
else {
 			sTicket = null;
	}
9. Type. Functions, General Programming: Don't Repeat Yourself (Avoid Duplication)
You have initialised the local variable sTicket to null:
else {
 			sTicket = null;
	}

**Review by duane-mcmahon — APPROVED** [review]
looks like it works!
I am not EXPECTING you to act on my comments.

**Comment by WBGroupR** [comment]
Thanks Duane, yes all good points - will do those changes. Just a git question - I guess I should pull down the remote to my master before doing those - as a matter of good practice. What do you think?

Also, Jim was suggesting that each separate change deserves a separate pull request (and thereby short review). Do you think we should do this or is it overkill given these are all really the same sort of mod (except the extraneous null)?

Wayne

File: src/bcccp/tickets/season/SeasonTicket.java (hunk 2 above)
Comment (WBGroupR) [reply src/bcccp/tickets/season/SeasonTicket.java:90]: ditto

File: src/bcccp/tickets/season/SeasonTicket.java (hunk 1 above)
Comment (WBGroupR) [reply src/bcccp/tickets/season/SeasonTicket.java:65]: Thanks Duane - good pickup

File: src/bcccp/tickets/season/SeasonTicketDAO.java (hunk 3 above)
Comment (WBGroupR) [reply src/bcccp/tickets/season/SeasonTicketDAO.java:49]: ditto

File: src/bcccp/tickets/season/SeasonTicketDAO.java (hunk 4 above)
Comment (WBGroupR) [reply src/bcccp/tickets/season/SeasonTicketDAO.java:79]: Yes, should have removed the null.

File: src/bcccp/tickets/season/SeasonTicket.java [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
+import java.util.Iterator;
 
 public class SeasonTicket implements ISeasonTicket {
```
Comment (duane-mcmahon) [inline src/bcccp/tickets/season/SeasonTicket.java:8]: 1. TYPE. Formatting: Google style Infringement
No javadocs present: "At the minimum, Javadoc is present for every public class,
 and every public or protected member of such a class".
2. TYPE. Fundamentals: Make class final if not being used for inheritance

**Comment by WBGroupR** [comment]
Thanks Duane, have now completed the fix and pushed new branch up to GitHub. Do you need to create the pull request or do I?

Wayne

File: src/bcccp/tickets/season/SeasonTicketDAO.java [hunk 6]
```diff
@@ -3,64 +3,108 @@
...
 
 
 	@Override
 	public void recordTicketExit(String ticketId) {
-		// TODO Auto-generated method stub
+
+		// Finds an existing usage record and records the current day and time (on exiting of vehicle)
+		// on the record
+
```
Comment (duane-mcmahon) [inline src/bcccp/tickets/season/SeasonTicketDAO.java:102]: 10. TYPE. Formatting: Google style Infringement
   Each time a new block or block-like construct is opened, the indent increases by two spaces. 
   This defect applies to the document as whole which appears to open with an indent of 4 spaces.

File: src/bcccp/tickets/season/SeasonTicketDAO.java [hunk 7]
```diff
@@ -3,64 +3,108 @@
 import bcccp.tickets.season.ISeasonTicket;
 import bcccp.tickets.season.IUsageRecordFactory;
 
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Date;
+
 public class SeasonTicketDAO implements ISeasonTicketDAO {
```
Comment (duane-mcmahon) [inline src/bcccp/tickets/season/SeasonTicketDAO.java:11]: 5. TYPE. Formatting: Google style Infringement
No javadocs present: "At the minimum, Javadoc is present for every public class,
and every public or protected member of such a class".

**Pull request merged**